io/slow5/slow5.go

```diff
@@ -0,0 +1,409 @@
+as a long string.
+
+Reads have 8 required columns, and a few auxillary. These are typed, since they
+are what will probably be used in real software.
+
+Cheers mate,
+
+Keoni
+
+******************************************************************************/
+
+// Header contains metadata about the sequencing run in general.
+type Header struct {
+	ReadGroupId  uint32
+	Slow5Version string
+	Attributes   map[string]string
+}
+
+// Read contains metadata and raw signal strengths for a single nanopore read.
+type Read struct {
+	ReadId       string
+	ReadGroupId  uint32
+	Digitisation float64
+	Offset       float64
+	Range        float64
+	SamplingRate float64
+	LenRawSignal uint64
+	RawSignal    []int16
+
+	// Auxiliary fields
+	ChannelNumber string
+	MedianBefore  float64
+	ReadNumber    int32
+	StartMux      uint8
+	StartTime     uint64
+	EndReason     string // enum{unknown,partial,mux_change,unblock_mux_change,data_service_unblock_mux_change,signal_positive,signal_negative}
+
+	Error error // in case there is an error while parsing!
+}
+
+var knownEndReasons = map[string]bool{"unknown": true,
+	"partial":                         true,
+	"mux_change":                      true,
+	"unblock_mux_change":              true,
+	"data_service_unblock_mux_change": true,
+	"signal_positive":                 true,
+	"signal_negative":                 true,
+}
+
+// Parser is a flexible parser that provides ample
+// control over reading slow5 sequences.
+// It is initialized with NewParser.
+type Parser struct {
+	// reader keeps state of current reader.
+	reader       bufio.Reader
+	line         uint
+	headerMap    map[int]string
+	endReasonMap map[int]string
+}
+
+// NewParser parsers a slow5 file.
+func NewParser(r io.Reader) (*Parser, []Header, error) {
+	const maxLineSize = 2 * 32 * 1024
+	parser := &Parser{
+		reader: *bufio.NewReaderSize(r, maxLineSize),
+		line:   0,
+	}
+	var headers []Header
+	var slow5Version string
+	var numReadGroups uint32
+	headerMap := make(map[int]string)
+	endReasonMap := make(map[int]string)
+
+	for {
+		lineBytes, err := parser.reader.ReadSlice('\n')
+		if err != nil {
+			return parser, []Header{}, err
+		}
+		line := strings.TrimSpace(string(lineBytes))
+		parser.line++
+		values := strings.Split(line, "\t")
+		if len(values) < 2 {
+			return parser, []Header{}, fmt.Errorf("Got following line without tabs: %s", line)
+		}
+
+		// First, we need to identify the number of read groups. This number will be the length of our
+		// ReadGroups output, and we will need it for iteration through the rest of the header.
+		if numReadGroups == 0 {
+			switch values[0] {
+			case "#slow5_version":
+				slow5Version = values[1]
+			case "#num_read_groups":
+				numReadGroupsUint, err := strconv.ParseUint(values[1], 10, 32)
+				if err != nil {
+					return parser, []Header{}, err
+				}
+				numReadGroups = uint32(numReadGroupsUint)
+				for id := uint32(0); id < numReadGroups; id++ {
+					emptyMap := make(map[string]string)
```

Comment:
You could inline emptyMap's creation with the append below. that is write `... Attributes: make(map[string]string)}`